tests/pso_test.cpp

```diff
@@ -0,0 +1,281 @@
+  upperBound.fill(4);
+
+  LBestPSO s(64, lowerBound, upperBound);
+
+  arma::mat coordinates = arma::mat("4.5; 4.5");
+  const double result = s.Optimize(f, coordinates);
+
+  REQUIRE(result == Approx(0).margin(0.01));
+  REQUIRE(coordinates[0] == Approx(3).margin(0.01));
+  REQUIRE(coordinates[1] == Approx(0.5).margin(0.01));
+}
+
+/**
+ * Test the PSO optimizer on the Goldstein-Price Function.
+ */
+TEST_CASE("LBestPSOGoldsteinPriceFunctionTest", "[PSOTest]")
+{
+  GoldsteinPriceFunction f;
+
+  // Setting bounds for the initial swarm population.
+  arma::vec lowerBound(2);
+  arma::vec upperBound(2);
+  lowerBound.fill(1.6);
+  upperBound.fill(2);
+
+  LBestPSO s(64, lowerBound, upperBound);
+
+  arma::mat coordinates = arma::mat("1; 0");
+  s.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.01));
+  REQUIRE(coordinates[1] == Approx(-1).margin(0.01));
+}
+
+/**
+ * Test the PSO optimizer on the Levi Function.
+ */
+TEST_CASE("LBestPSOLevyFunctionN13Test", "[PSOTest]")
+{
+  LevyFunctionN13 f;
+
+  // Setting bounds for the initial swarm population.
+  arma::vec lowerBound(2);
+  arma::vec upperBound(2);
+  lowerBound.fill(-10);
+  upperBound.fill(-9);
+
+  LBestPSO s(64, lowerBound, upperBound);
+
+  arma::mat coordinates = arma::mat("3; 3");
+  s.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(1).margin(0.01));
+  REQUIRE(coordinates[1] == Approx(1).margin(0.01));
+}
+
+/**
+ * Test the PSO optimizer on the Himmelblau Function.
+ */
+TEST_CASE("LBestPSOHimmelblauFunctionTest", "[PSOTest]")
+{
+  HimmelblauFunction f;
+
+  // Setting bounds for the initial swarm population.
+  arma::vec lowerBound(2);
+  arma::vec upperBound(2);
+  lowerBound.fill(0);
+  upperBound.fill(1);
+
+  LBestPSO s(64, lowerBound, upperBound);
+
+  arma::mat coordinates = arma::mat("2; 1");
+  s.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(3.0).margin(0.01));
+  REQUIRE(coordinates[1] == Approx(2.0).margin(0.01));
+}
+
+/**
+ * Test the PSO optimizer on the Three-hump Camel Function.
+ */
+TEST_CASE("LBestPSOThreeHumpCamelFunctionTest", "[PSOTest]")
+{
+  ThreeHumpCamelFunction f;
+
+  // Setting bounds for the initial swarm population.
+  arma::vec lowerBound(2);
+  arma::vec upperBound(2);
+  lowerBound.fill(-5);
+  upperBound.fill(-4);
+
+  LBestPSO s(64, lowerBound, upperBound);
+
+  arma::mat coordinates = arma::mat("2; 2");
+  s.Optimize(f, coordinates);
+
+  REQUIRE(coordinates[0] == Approx(0).margin(0.01));
+  REQUIRE(coordinates[1] == Approx(0).margin(0.01));
+}
+
+/**
+ * Test the PSO optimizer on Schaffer function N.4.
+ */
+TEST_CASE("LBestPSOScafferFunctionN4Test", "[PSOTest]")
+{
+  SchafferFunctionN4 f;
+ 
+  // Setting bounds for the initial swarm population.
+  arma::vec lowerBound(2);
+  arma::vec upperBound(2);
+  lowerBound.fill(40);
+  upperBound.fill(50);
+  
+  LBestPSO s(15000, lowerBound, upperBound, 3000, 200, 1e-20, 1.5, 3.0);
+  arma::mat coordinates = arma::mat("0; 10");
```

Comment:
Perhaps it makes sense to use an initial starting point that is closer to the solution.

Reply:
That's the one thing I did not want to do. I wanted to test PSO for its robustness with non-convex problems. As a dynamical system, I'm not sure if selecting a closer initial point in a problem this hard will make much of a difference. Regardless, let me try this out. If it works, then we'll have a new observation. Also, let me try to initialize the solutions over a larger area that encompasses the solution. That's one thing that we haven't tried.

Reply:
I've been trying for the past two hours with many different values. Can't seem to land a 100% success rate. If you're okay with it, may we exclude it for the time being?

Reply:
Sure, let's comment the test for now.